File: python/lesson 9/course_schedule.py
# https://leetcode.com/problems/course-schedule-ii/

class DirectedGraph:
    def __init__(self, capacity: int):
        self.adj = [[]] * (capacity)
        for i in range(0, capacity-1):
            self.adj[i] = []
        
    def size(self) -> int:
        return len(self.adj)
    
    def get(self, v: int) -> List[int]:
        return self.adj[v]
    
    def add(self, v: int, w: int):
        self.adj[v].append(w)
        
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = DirectedGraph(numCourses)
        for p in prerequisites:
            graph.add(p[1], p[0])
        return self.topologicalSort(graph)
    
    def topologicalSort(self, graph: DirectedGraph) -> List[int]:
        visited = [False] * (graph.size())
        marked = [False] * (graph.size())
        stack = []

        for v in range(0, graph.size()):
            if not self.dfs(graph, v, visited, marked, stack): return []
            
        return reversed(stack)
            
    def dfs(self, graph: DirectedGraph, v: int, visited: List[bool], marked: List[bool], stack: List[int]) -> bool:
        if marked[v]: return False
        if visited[v]: return True
        
        visited[v] = True
        marked[v] = True
        
        for w in graph.get(v):
            if not self.dfs(graph, w, visited, marked, stack): return False

        stack.append(v)
        marked[v] = False
        
        return True

File: kotlin/lesson 9/CourseSchedule.kt
// https://leetcode.com/problems/course-schedule-ii/

class Solution {
    fun findOrder(numCourses: Int, prerequisites: Array<IntArray>): IntArray {
        val graph = DirectedGraph(numCourses)
        for (p in prerequisites) {
            graph.add(p[1], p[0])
        }
        return topologicalSort(graph)
    }
    
    fun topologicalSort(graph: DirectedGraph): IntArray {
        val stack = Stack<Int>()
        val visited = BooleanArray(graph.size())
        val marked = BooleanArray(graph.size())
        
        for (v in 0..graph.size() - 1) {
            if (!dfs(graph, v, visited, marked, stack)) return IntArray(0)
        }
        
        val result = IntArray(stack.size);
        var i = 0
        while (!stack.isEmpty()) {
            result[i] = stack.pop()
            i++
        }
        return result
    }
    
    fun dfs(graph: DirectedGraph, v: Int, visited: BooleanArray, marked: BooleanArray, stack: Stack<Int>): Boolean {
        if (marked[v]) return false
        if (visited[v]) return true
        
        marked[v] = true
        visited[v] = true
                
        for (w in graph.get(v)) {
            if (!dfs(graph, w, visited, marked, stack)) return false
        }
        
        stack.push(v)
        marked[v] = false
        
        return true
    }
    
    public class DirectedGraph(capacity: Int) {
        val adj = Array<LinkedList<Int>>(capacity)  { LinkedList<Int>() }

        fun get(v: Int): List<Int> = adj[v]

        fun add(v: Int, w: Int) {
            adj[v].add(w)
        }

        fun size() = adj.size
    }
}
